### DISCUSSION TITLE
Using Route Groups for Route Handlers

### DISCUSSION THREAD

**Discussion by jonashoen** [vercel/next.js#54178] [App Router]
I was using route groups to group my pages into authed and un-authed ones. When using this approach for route handlers, and possibly server actions, one could create middlewares that are only run when accessing a specific route segment. For example, to split every route into authed and un-authed ones as well, or to create a custom middleware to check for admin privileges.

I personally don't like the current middleware with its path matching as it introduces configuration which would need to be extended every time one adds another route. If instead the new route just gets created in the same route groups, no configuration would be needed.

The file structure would then look something like this:

![image](https://github.com/vercel/next.js/assets/56582545/3dd5b133-b540-4ced-8fff-3d88ea0cd327)

The feature could be implemented using the existing `NextResponse.next()`. It would run before every route inside the route group is executed so that the calls could either be passed to the corresponding endpoint or be aborted with some error code.

**Comment by zetorama** [comment]
Unless i'm missing something, this could also help a lot with running common code for multiple route handlers.

For example, it's easy to call `initSomething()` in any `layout.ts`, expecting that all `page.ts`s within this layout are getting that. But it seems that currently for `route.ts`s there's no way to do anything similar, so each route handles should have its own `initSomething()` call.

**Comment by jonashoen** [comment]
Yeah exactly. Currently I'm using an express-like middleware approach by exporting my route handler function wrapped in a middleware function:

```ts
const POST = withMiddleware([middleware1, middleware2, ...], handler);
export { POST };
```

But I feel like integrating this into the file system would be a better fit in the nextjs architecture.
